File: program/programs/clad/src/instructions/create_pool.rs
use {
    crate::state::*,
    anchor_lang::prelude::*,
    anchor_spl::token::{self, Mint, Token, TokenAccount},
};

#[derive(Accounts)]
#[instruction(params: CreatePoolParams)]
pub struct CreatePool<'info> {
    #[account(mut)]
    pub funder: Signer<'info>,

    #[account(
        seeds = [b"clad"],
        bump = clad.clad_bump,
    )]
    pub clad: Box<Account<'info, Clad>>,

    #[account(
        init,
        seeds = [
            b"globalpool".as_ref(),
            params.token_mint_a.key().as_ref(),
            params.token_mint_b.key().as_ref(),
            params.fee_rate.to_le_bytes().as_ref(),
            params.tick_spacing.to_le_bytes().as_ref(),
        ],
        bump,
        payer = funder,
        space = Globalpool::LEN
    )]
    pub globalpool: Box<Account<'info, Globalpool>>,

    // CHECK: empty PDA, will be set as authority for token accounts
    // #[account(
    //     init,
    //     payer = funder,
    //     space = 0,
    //     seeds = [
    //         b"transfer_authority",
    //         globalpool.key().as_ref()
    //     ],
    //     bump
    // )]
    // pub transfer_authority: AccountInfo<'info>,
    pub token_mint_a: Account<'info, Mint>,

    pub token_mint_b: Account<'info, Mint>,

    #[account(
        init,
        payer = funder,
        token::mint = token_mint_a,
        token::authority = globalpool // transfer_authority
    )]
    pub token_vault_a: Box<Account<'info, TokenAccount>>,

    #[account(
        init,
        payer = funder,
        token::mint = token_mint_b,
        token::authority = globalpool // transfer_authority
    )]
    pub token_vault_b: Box<Account<'info, TokenAccount>>,

    #[account(address = token::ID)]
    pub token_program: Program<'info, Token>,

    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
pub struct CreatePoolParams {
    pub fee_rate: u16,
    pub tick_spacing: u16,
    pub initial_sqrt_price: u128,
}

pub fn create_pool(ctx: Context<CreatePool>, params: &CreatePoolParams) -> Result<()> {
    let clad = ctx.accounts.clad;

    let globalpool = &mut ctx.accounts.globalpool;
    let globalpool_bump = *ctx
        .bumps
        .get("globalpool")
        .ok_or(ProgramError::InvalidSeeds)?;
    let transfer_authority_bump = *ctx
        .bumps
        .get("transfer_authority")
        .ok_or(ProgramError::InvalidSeeds)?;

    Ok(globalpool.initialize(
        globalpool_bump,
        transfer_authority_bump,
        params.tick_spacing,
        params.initial_sqrt_price,
        params.fee_rate,
        clad.protocol_fee_rate,
        ctx.accounts.token_mint_a.key(),
        ctx.accounts.token_vault_a.key(),
        ctx.accounts.token_mint_b.key(),
        ctx.accounts.token_vault_b.key(),
    )?)
}

File: scripts/src/actions/create-pool.ts
import * as anchor from '@coral-xyz/anchor'
import { TransactionBuilder } from '@orca-so/common-sdk'
import { PriceMath, TICK_ARRAY_SIZE, TickUtil } from '@orca-so/whirlpools-sdk'
import {
  ASSOCIATED_TOKEN_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
  getAssociatedTokenAddressSync,
} from '@solana/spl-token'
import {
  Connection,
  PublicKey,
  SYSVAR_RENT_PUBKEY,
  SystemProgram,
} from '@solana/web3.js'
import Decimal from 'decimal.js'

import { Clad } from '../constants'
import { getConstantParams } from '../params'
import poolConfigs from '../pool-config.json'
import { ParsableGlobalpool } from '../types/parsing'
import { getAccountData } from '../utils'
import { getPythPrices } from '../utils/pyth'
import { createTransactionChained } from '../utils/txix'
import { initTickArrayRange } from '../utils/tick-arrays'

async function createPools(connection: Connection, program: anchor.Program<Clad>, wallet: anchor.Wallet, provider: anchor.AnchorProvider, cladKey: PublicKey) {
  const symbolSets = new Set<string>()

  Object.values(poolConfigs).map((poolConfig) => {
    symbolSets.add(poolConfig.baseToken.symbol)
    symbolSets.add(poolConfig.quoteToken.symbol)
  })

  const prices = await getPythPrices([...symbolSets])

  const initPromises = Object.values(poolConfigs).map(async (poolConfig) => {
    const tokenMintAKey = new PublicKey(poolConfig.baseToken.address)
    const tokenMintBKey = new PublicKey(poolConfig.quoteToken.address)

    // const tokenMintA = await getMint(connection, tokenMintAKey)
    // const tokenMintB = await getMint(connection, tokenMintBKey)

    const decimalsA = poolConfig.baseToken.decimals
    const decimalsB = poolConfig.quoteToken.decimals

    const tokenPriceA = prices[poolConfig.baseToken.symbol]
    const tokenPriceB = prices[poolConfig.quoteToken.symbol]

    if (!tokenPriceA || !tokenPriceB) {
      console.error(`Token price not found for ${poolConfig.baseToken.symbol} or ${poolConfig.quoteToken.symbol}`)
      return
    }

    const desiredStartPrice = new Decimal(tokenPriceA).div(tokenPriceB)
    const tickSpacing = poolConfig.tickSpacing
    const feeRate = poolConfig.feeRate

    const initTickIndex = PriceMath.priceToTickIndex(desiredStartPrice, decimalsA, decimalsB) // Math.round((Math.log(desiredStartPrice * Math.pow(10, decimalDiff)) / Math.log(1.0001)) / tickSpacing) * tickSpacing
    const initPrice = PriceMath.tickIndexToPrice(initTickIndex, decimalsA, decimalsB)
    const initSqrtPrice = PriceMath.tickIndexToSqrtPriceX64(initTickIndex)

    const globalpoolSeeds = [
      Buffer.from('globalpool'),
      tokenMintAKey.toBuffer(),
      tokenMintBKey.toBuffer(),
      new anchor.BN(feeRate).toArrayLike(Buffer, 'le', 2),
      new anchor.BN(tickSpacing).toArrayLike(Buffer, 'le', 2),
    ]

    const [globalpoolKey] = PublicKey.findProgramAddressSync(
      globalpoolSeeds,
      program.programId
    )

    const globalpoolInfo = await getAccountData(globalpoolKey, ParsableGlobalpool, connection)
    if (globalpoolInfo) {
      console.log(`Globalpool ${globalpoolKey.toBase58()} already exists`)
      return
    }

    const tokenVaultA = getAssociatedTokenAddressSync(tokenMintAKey, globalpoolKey, true)
    const tokenVaultB = getAssociatedTokenAddressSync(tokenMintBKey, globalpoolKey, true)

    const createPoolParams = {
      feeRate,
      tickSpacing,
      initialSqrtPrice: initSqrtPrice,
    }

    const createPoolAccounts = {
      funder: wallet.publicKey,
      clad: cladKey,
      globalpool: globalpoolKey,
      tokenMintA: tokenMintAKey,
      tokenMintB: tokenMintBKey,
      tokenVaultA: tokenVaultA,
      tokenVaultB: tokenVaultB,
      // sys
      tokenProgram: TOKEN_PROGRAM_ID,
      associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
      systemProgram: SystemProgram.programId,
      rent: SYSVAR_RENT_PUBKEY,
    }

    await createTransactionChained(
      connection,
      wallet,
      program.instruction.createPool(createPoolParams, {
        accounts: createPoolAccounts,
      }),
      []
    ).buildAndExecute()

    console.log('Created Pool: ', globalpoolKey.toString())
    console.log('   Base:  ', poolConfig.baseToken.symbol)
    console.log('   Quote: ', poolConfig.quoteToken.symbol)
    console.log('   Fee Rate:      ', feeRate)
    console.log('   Tick Spacing:  ', tickSpacing)
    console.log('   Initial Price: ', initPrice)
    console.log('   Initial Tick:  ', initTickIndex)

    await initializePoolTicks(globalpoolKey, initTickIndex, tickSpacing, program, provider)

    return Promise.resolve(true)
  })

  await Promise.all(initPromises)
}

async function initializePoolTicks(globalpoolKey: PublicKey, initTickIndex: number, tickSpacing: number, program: anchor.Program<Clad>, provider: anchor.AnchorProvider) {
  const aToB = false // determines direction of tick array

  const initArrayCount = 17 // 8 to left of, 8 to right of, and 1 array containing current tick
  const currentTickArrayStartIndex = TickUtil.getStartTickIndex(
    initTickIndex,
    tickSpacing
  )

  // reverse direction of `aToB` because `initTickArrayRange` will init in the direction
  // of `aToB` (left if false, right if true)
  const startTickIndex =
    currentTickArrayStartIndex +
    (aToB ? 1 : -1) *
    Math.floor(initArrayCount / 2) *
    tickSpacing *
    TICK_ARRAY_SIZE

  await initTickArrayRange(
    globalpoolKey,
    startTickIndex,
    initArrayCount,
    tickSpacing,
    aToB,
    program,
    provider
  )

  console.log('Initialized some Tick Arrays around the current tick')
}

async function main() {
  const {
    program,
    connection,
    provider,
    wallet,
    cladKey,
  } = await getConstantParams()

  //
  // Create Clad account
  //

  const cladAccountInfo = await connection.getAccountInfo(cladKey)

  if (!cladAccountInfo) {
    const initCladParams = {
      permissions: {
        allowSwap: true,
        allowAddLiquidity: true,
        allowRemoveLiquidity: true,
        allowOpenPosition: true,
        allowClosePosition: true,
      },
      protocolFeeRate: 0,
    }
    const initCladAccounts = {
      admin: wallet.publicKey,
      clad: cladKey,
      systemProgram: SystemProgram.programId,
      rent: SYSVAR_RENT_PUBKEY,
    }

    const initCladTx = new TransactionBuilder(connection, wallet)
      .addInstruction({
        instructions: [
          program.instruction.initializeClad(initCladParams, {
            accounts: initCladAccounts,
          }),
        ],
        cleanupInstructions: [],
        signers: [],
      })

    await initCladTx.buildAndExecute()

    console.log('Initialized Clad: ', cladKey.toBase58())
  } else {
    console.log('Clad already exists: ', cladKey.toBase58())
  }

  //
  // Create Pools & initialize some tick arrays
  //
  await createPools(connection, program, wallet as anchor.Wallet, provider, cladKey)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
